### ISSUE TITLE
ft qwen2的时候，flash attn 和core attn的输出相差较大，且attn_mask为false的token, flash attn输出的是全0向量，但core attn输出的是一个正常向量

### ISSUE THREAD

**Issue by seanM29** [QwenLM/Qwen2.5#874]
训练的时候，发现启用不同的attn implement, 对loss有比较大的影响

定位发现，同样的q,k,v 输入值，flash attn 和core attn的输出相差较大，

且attn_mask为false的token, flash attn输出的是全0向量，但core attn输出的是一个正常向量, 

不算这个0向量的话，flash attn和core attn输出的output token 相似度低的也有到0.8左右，不知道是否符合预期？感谢！

**Comment by jklj077** [comment]
what is core attn?

**Comment by seanM29** [comment]
感谢你的回复，core attn是eager attn，不好意思没有表达清楚

我是实验发现，eager attn的loss和性能相比flash_attn差异很大，所以去定位排查

0向量的问题：
1）flash attn输入的attn_mask是bs*seq的2d tensor，假如attn_mask[i,j]=False，即认为这个token是pad token，则flash attn输出的attn_output（shape为bs*seq*hidden_dim）,attn_output[i,j]为zero向量，然而eager attn输出的attn_output[i,j]为非0向量
2）不过后来去仔细看了，因为eager的attn mask会让这个padtoken对其他token没有影响，且这个pad token不产生loss，所以其实还好？不知道这个输出是否符合预期？


除此以外，我其他token的attn_output 相似度比较低的话，会到0.8，这个其实感觉影响蛮大的？

另外我会整理tensor和权重来稳定复现这个问题，最后感谢你的回复，谢谢！

**Comment by jklj077** [comment]
Hello,

Thank you for your clarification. It appears that the zero tensors produced by `flash_attn` are indeed expected when dealing with padding, as `flash_attn` skips the computation of padding tokens altogether. In contrast, the eager attention implementation in `transformers` does compute the padding tokens, but they are masked out during the attention computation.[^eager] Since the padding tokens are also masked out in the loss computation, the observed differences between the two methods are likely insignificant.

Could you share a bit more information about the accuracy differences you've noticed?

- Was the result with `flash_attn` better, or was the one with the eager implementation better?
- Which model did you use for fine-tuning?
- What similarity metric did you use? Was it cosine similarity?
- How about the mean squared difference?

These details would be quite helpful.

[^eager]: The eager attention implementation in `transformers` is quite versatile. It can accept a 4D attention mask (with dimensions `[batch_size, heads, target_sequence_length, source_sequence_length]`), allowing for more nuanced attention strategies. However, for causal language models, we typically pass a 2D attention mask that indicates padding, and this is transformed into a causal mask (a lower triangular matrix) within the implementation.

**Comment by seanM29** [comment]
再次感谢你的回复

1. loss的话，eager attn更低，但是验证集的评测指标flash attn更好，sdpa的loss和指标和eager差不多，同时我观察到eager attn更容易出现loss 波动，所有都是在bf16下训练的
2. qwen2-7b
3. cosine similarity
4. l2距离的话，attn_mask false的pad token 在6-8，排除pad token后，最大的距离在5.4，所有token平均距离是2.5

**Comment by jklj077** [comment]
Hi, for Qwen2-7B, I would say it is expected and the difference should be mainly caused by issues with numerical stability.

**Comment by seanM29** [comment]
非常感谢你的回复，因为我要得倒attn中的attn_weights，也就是 output_attentions是true，所以必须要使用eager attn

一开始也怀疑过数值稳定性，因为eager attn有一次nan了，就使用了sdpa attn，loss曲线和eager attn基本重合，但是训练非常稳定，同时loss 比flash attn低，这个蛮奇怪的

如果是数值稳定性的话，请问有什么建议嘛？如果我要得倒attn_weights的话？用eager attn 然后使用fp32? 降低lr？

 再次谢谢你的回复了！

**Comment by jklj077** [comment]
To use eager implementation with 7B in _inference_, please refer to [this](https://github.com/jklj077/transformers/blob/53dba62c9c35c4a632c81f0406d844ea3679d4dd/src/transformers/models/qwen2/modeling_qwen2.py#L285-L289). 

But since you are finetuning and bfloat16 is being used, I don't think that specific hack would work. Using float32 and lowering learning rate or applying a strong L2 norm to attention weights all appear reasonable.

**Comment by seanM29** [comment]
> To use eager implementation with 7B in _inference_, please refer to [this](https://github.com/jklj077/transformers/blob/53dba62c9c35c4a632c81f0406d844ea3679d4dd/src/transformers/models/qwen2/modeling_qwen2.py#L285-L289).
> 
> But since you are finetuning and bfloat16 is being used, I don't think that specific hack would work. Using float32 and lowering learning rate or applying a strong L2 norm to attention weights all appear reasonable.

@jklj077 感谢您的回复和时间，这个看起来就是qwen2 的model py？我目前使用的就是这个eager attention版本，将softmax转为fp32

能多问下，“a strong L2 norm to attention weights”指的是什么吗？是对softmax过后的attn_weights增加一个l2的正则约束吗？为什么要这么做呢？L2惩罚了极大值，这是有什么相关文献嘛？ 

谢谢了！
